### PR TITLE
[RFC] make del_var and set_var not return the old value.

### PR DESCRIPTION
**Pull request by bfredl**
As discussed in #4411. 
#4568 could be used emulate the old behavior, or any client that relies on the old behavior could continue to use the deprecated functions for now; we are not in a hurry removing deprecated functions.

@ZyX-I sorry for duplicating the change in #5119 where `deprectated_dispatch.lua` is moved to the right location. I needed to fix the cmake rule  (missing dependency) and also wanted to mention the name of the file in an error message.

### CODE REVIEW

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -432,19 +432,52 @@ Object nvim_buf_get_var(Buffer buffer, String name, Error *err)
...
+/// @param buffer The buffer handle
+/// @param name The variable name
+/// @param value The variable value
+/// @param[out] err Details of an error that may have occurred
 /// @return The old value or nil if there was no previous value.
 ///
 ///         @warning It may return nil if there was no previous value
 ///                  or if previous value was `v:null`.
-Object nvim_buf_set_var(Buffer buffer, String name, Object value, Error *err)
+Object buffer_set_var(Buffer buffer, String name, Object value, Error *err)
```
Comment (justinmk) [inline src/nvim/api/buffer.c]: mark `@deprecated` ?

File: src/nvim/api/private/helpers.c [hunk 2]
```diff
@@ -104,10 +104,12 @@ Object dict_get_value(dict_T *dict, String key, Error *err)
 /// @param value The new value
 /// @param del Delete key in place of setting it. Argument `value` is ignored in
 ///            this case.
+/// @param retval If true the old value will be converted and returned. Otherwise NIL is 
+///            this case.
 /// @param[out] err Details of an error that may have occurred
 /// @return the old value, if any
```
Comment (justinmk) [inline src/nvim/api/private/helpers.c]: @return old value if `reval` is true, else NIL

File: src/nvim/api/private/helpers.c [hunk 3]
```diff
@@ -104,10 +104,12 @@ Object dict_get_value(dict_T *dict, String key, Error *err)
 /// @param value The new value
 /// @param del Delete key in place of setting it. Argument `value` is ignored in
 ///            this case.
+/// @param retval If true the old value will be converted and returned. Otherwise NIL is 
+///            this case.
```
Comment (justinmk) [inline src/nvim/api/private/helpers.c]: typo: "NIL in" (but I suggest omitting that last sentence, let `@return` cover the explanation)

File: src/nvim/api/buffer.c [hunk 4]
```diff
@@ -453,20 +486,18 @@ Object nvim_buf_set_var(Buffer buffer, String name, Object value, Error *err)
 /// @param name The variable name
 /// @param[out] err Details of an error that may have occurred
 /// @return The old value or nil if there was no previous value.
-///
-///         @warning It may return nil if there was no previous value
-///                  or if previous value was `v:null`.
-Object nvim_buf_del_var(Buffer buffer, String name, Error *err)
+Object buffer_del_var(Buffer buffer, String name, Error *err)
```
Comment (justinmk) [inline src/nvim/api/buffer.c]: is the `@return` doc correct?

File: src/nvim/api/buffer.c (hunk 1 above)
Comment (bfredl) [inline src/nvim/api/buffer.c]: I thought about it, but ain't functions with these names implicitly deprecated? When implementing the doc generation the plan is to not generate docs for these, or alternatively place them in a "deprecated" section. But anyway, it shouldn't hurt to be extra clear, so I'll add it.

File: src/nvim/api/buffer.c (hunk 4 above)
Comment (bfredl) [inline src/nvim/api/buffer.c]: No, it is only the old value, if key is not found it is an error. Thanks for noting!

**Comment by bfredl** [comment]
@justinmk Thanks for the comments, updated.

**Comment by justinmk** [comment]
LGTM

File: test/functional/api/vim_spec.lua [hunk 5]
```diff
@@ -88,19 +88,19 @@ describe('vim_* functions', function()
...
+    it('vim_del_var returns the old value', function()
       local val1 = {1, 2, {['3'] = 1}}
       local val2 = {4, 7}
-      eq(NIL, meths.set_var('lua', val1))
-      eq(val1, meths.set_var('lua', val2))
-      eq(val2, meths.del_var('lua'))
+      eq(NIL,  request('vim_set_var', 'lua', val1))
+      eq(val1, request('vim_set_var', 'lua', val2))
+      eq(val2, request('vim_del_var', 'lua'))
     end)
```
Comment (ZyX-I) [inline test/functional/api/vim_spec.lua]: Where are tests for *new* functions? Also AFAIR all `*_set_var` and `*_del_var` are tested, not just `vim_*`.

Comment (bfredl) [inline test/functional/api/vim_spec.lua]: The one implementing this to start with apparently thought it was redundant, but adding for the others should be a quick matter of find and replace

**Pull request merged**